File: examples/python/agg-order/handlers/cancel_order.py
"""Handler for CancelOrder command."""

from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import types_pb2 as types
from errors import CommandRejectedError
from proto import order_pb2 as order

from .state import OrderState


def handle_cancel_order(
    command_book: types.CommandBook,
    command_any: Any,
    state: OrderState,
    seq: int,
) -> types.EventBook:
    """Handle CancelOrder command."""
    if not state.exists():
        raise CommandRejectedError("Order does not exist")
    if state.is_completed():
        raise CommandRejectedError("Cannot cancel completed order")
    if state.is_cancelled():
        raise CommandRejectedError("Order already cancelled")

    cmd = order.CancelOrder()
    command_any.Unpack(cmd)

    if not cmd.reason:
        raise CommandRejectedError("Cancellation reason is required")

    event = order.OrderCancelled(
        reason=cmd.reason,
        cancelled_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
        loyalty_points_used=state.loyalty_points_used,
        customer_root=state.customer_root,
        cart_root=state.cart_root,
    )
    event.items.extend(
        order.LineItem(
            product_id=i.product_id,
            name=i.name,
            quantity=i.quantity,
            unit_price_cents=i.unit_price_cents,
        )
        for i in state.items
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return types.EventBook(
        cover=command_book.cover,
        pages=[
            types.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
            )
        ],
    )

File: examples/rust/agg-order/src/handlers/cancel_order.rs
//! Handler for CancelOrder command.

use angzarr::proto::{CommandBook, EventBook};
use common::proto::{CancelOrder, OrderCancelled, OrderState};
use common::{decode_command, now, require_exists, require_status_not, ProtoTypeName, Result};

use crate::errmsg;
use crate::state::state_builder;
use crate::status::OrderStatus;

/// Handle the CancelOrder command.
pub fn handle_cancel_order(
    command_book: &CommandBook,
    command_data: &[u8],
    state: &OrderState,
    next_seq: u32,
) -> Result<EventBook> {
    require_exists(&state.customer_id, errmsg::ORDER_NOT_FOUND)?;
    require_status_not(
        &state.status,
        OrderStatus::Completed.as_str(),
        errmsg::ORDER_COMPLETED,
    )?;
    require_status_not(
        &state.status,
        OrderStatus::Cancelled.as_str(),
        errmsg::ORDER_CANCELLED,
    )?;

    let cmd: CancelOrder = decode_command(command_data)?;

    let event = OrderCancelled {
        reason: cmd.reason,
        cancelled_at: Some(now()),
        loyalty_points_used: state.loyalty_points_used,
        customer_root: state.customer_root.clone(),
        items: state.items.clone(),
        cart_root: state.cart_root.clone(),
    };

    Ok(state_builder().build_response(
        state,
        command_book.cover.clone(),
        next_seq,
        &OrderCancelled::type_url(),
        event,
    ))
}
